storage/operation/stats.go

```diff
@@ -0,0 +1,182 @@
+package operation
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/module/util"
+	"github.com/onflow/flow-go/storage"
+)
+
+// Stats holds statistics for a single prefix group.
+type Stats struct {
+	Count       int     `json:"count"`
+	MinSize     int     `json:"min_size"`
+	MaxSize     int     `json:"max_size"`
+	TotalSize   int     `json:"total_size"`
+	AverageSize float64 `json:"avg_size"`
+}
+
+// SummarizeKeysByFirstByteConcurrent iterates over all prefixes [0x00..0xFF] in parallel
+// using nWorker goroutines. Each worker handles one prefix at a time until all are processed.
+//
+// The storage.Reader must be able to create multiple iterators concurrently.
+func SummarizeKeysByFirstByteConcurrent(log zerolog.Logger, r storage.Reader, nWorker int) (map[byte]Stats, error) {
+	// We'll have at most 256 possible prefixes (0x00..0xFF).
+	// Create tasks (one per prefix), a results channel, and a wait group.
+	taskChan := make(chan byte, 256)
+	resultChan := make(chan struct {
+		prefix byte
+		stats  Stats
+		err    error
+	}, 256)
+
+	var wg sync.WaitGroup
+
+	// Start nWorker goroutines.
+	for i := 0; i < nWorker; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for prefix := range taskChan {
+				st, err := processPrefix(r, prefix)
+				resultChan <- struct {
+					prefix byte
+					stats  Stats
+					err    error
+				}{
+					prefix: prefix,
+					stats:  st,
+					err:    err,
+				}
+			}
+		}()
+	}
+
+	progress := util.LogProgress(log,
+		util.DefaultLogProgressConfig(
+			"Summarizing keys by first byte",
+			256,
+		))
+
+	// Send all prefixes [0..255] to taskChan.
+	for p := 0; p < 256; p++ {
+		taskChan <- byte(p)
+	}
+	close(taskChan)
+
+	// Once all workers finish, close the result channel.
+	go func() {
+		wg.Wait()
+		close(resultChan)
+	}()
+
+	// Gather results. We'll accumulate them in a map[prefix]Stats.
+	finalStats := make(map[byte]Stats, 256)
+
+	// If we encounter an error, we will return it immediately.
+	for res := range resultChan {
+		if res.err != nil {
+			return nil, res.err
+		}
+		finalStats[res.prefix] = res.stats
+		log.Info().
+			Int("prefix", int(res.prefix)).
+			Int("total", res.stats.TotalSize).
+			Int("count", res.stats.Count).
+			Int("min", res.stats.MinSize).
+			Int("max", res.stats.MaxSize).
+			Msg("Processed prefix")
+		progress(1) // log the progress
+	}
+
+	return finalStats, nil
```

Comment:
To prevent running goroutines reading from db after this function returns on error, this function needs to cancel running goroutines on error and waits for them to complete before returning error.

```suggestion
	var err error

	// If we encounter an error, we will return it immediately.
	for res := range resultChan {
		if res.err != nil {
			cancel() // Cancel running goroutines
			err = res.err
			break 
		}
		finalStats[res.prefix] = res.stats
		log.Info().
			Int("prefix", int(res.prefix)).
			Int("total", res.stats.TotalSize).
			Int("count", res.stats.Count).
			Int("min", res.stats.MinSize).
			Int("max", res.stats.MaxSize).
			Msg("Processed prefix")
		progress(1) // log the progress
	}

	<-done // Wait for worker goroutines
	if err != nil {
		return nil, err
	}
	return finalStats, nil
```